File: LeetCode/Medium/0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
    dummy := &Node{}
    temp := dummy
    ret := root
    
    for root != nil {
        if root.Left != nil {
            temp.Next = root.Left
            temp = temp.Next
        }
        
        if root.Right != nil {
            temp.Next = root.Right
            temp = temp.Next
        }
        
        root = root.Next
        
        if root == nil {
            root = dummy.Next
            dummy.Next = nil
            temp = dummy
        }
    }
    
    return ret
}

File: LeetCode/Medium/0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        Node* dummy = new Node();
        Node* temp = dummy;
        Node* ret = root;
        
        while (root){
            if (root->left){
                temp->next = root->left;
                temp = temp->next;
            }
            
            if (root->right){
                temp->next = root->right;
                temp = temp->next;
            }
            
            root = root->next;
            
            if (!root){
                root = dummy->next;
                dummy->next = nullptr;
                temp = dummy;
            }
        }
        
        return ret;
    }
};